applications/RomApplication/custom_utilities/rom_auxiliary_utilities.h

```diff
@@ -266,23 +272,45 @@ class KRATOS_API(ROM_APPLICATION) RomAuxiliaryUtilities
         const std::unordered_map<Kratos::VariableData::KeyType, Matrix::size_type>& rVarToRowMapping);
 
     /**
-     * @brief Obtain the JPhi elemental matrix for a particular element. 
+     * @brief Obtain the JPhi elemental matrix for a particular element.
      * JPhi represents the projection of the Jacobian onto the ROM_BASIS.
      * @param rJPhiElemental The matrix to store the result in. Must have the appropriate size already.
      * @param rDofs The set of degrees of freedom (DoFs) of the element.
      * @param rJPhi The JPhi matrix, from which rows are extracted according to the equation ID of each DoF.
-     * 
-     * This function loops over all the DoFs for the given element. For each DoF, it uses its equation ID to extract a 
+     *
+     * This function loops over all the DoFs for the given element. For each DoF, it uses its equation ID to extract a
      * corresponding row from the rJPhi matrix, which is then stored in the corresponding row of rJPhiElemental.
      */
     static void GetJPhiElemental(
         Matrix &rJPhiElemental,
         const Element::DofsVectorType& rDofs,
         const Matrix &rJPhi);
-        
+
+    /**
+     * @brief Finds the nearest neighbors for a set of nodes in a given model part.
+     *
+     * This function locates the closest node in the master structure to each node in the provided vector.
+     * It uses a spatial bin to efficiently search for the nearest neighbor. The IDs of the nearest nodes are
+     * stored in an unordered set to ensure uniqueness.
+     *
+     * @param rMasterStructureNodes Reference to the master structure nodes, which contain all the nodes in the model part.
+     * @param nodesVector A vector of pointers to the nodes for which the nearest neighbors are to be found.
+     * @return std::unordered_set<int> A set containing unique IDs of the nearest nodes.
+     */
+    std::unordered_set<int> FindNearestNeighbors(
+        NodesPointerSetType& rMasterStructureNodes,
+        std::vector<Node::Pointer> nodesVector);
+
     ///@}
 
     private:
+        ///@}
+        ///@name Member Variables
+        ///@{
+
+        NodeBinsType::UniquePointer mpBins;
```

Comment:
`mpBins` appears to be used only in the new `FindNearestNeighbors` function and its regenerated every time the function is called. 

If this is the case please just remove the variable and declare it inside the function.

If you intend the utility to be reusable (calculate the bins once) then please move the initialization of the member variable to a separate function and then use the pointer.